src/physics/linear_filament.rs

```diff
@@ -216,82 +199,96 @@ pub fn flux_density_linear_filament(
     // the compiler will bypass vectorization
     let n = xfil.len();
     let m = xp.len();
-
-    if xp.len() != m
-        || yp.len() != m
-        || zp.len() != m
-        || xfil.len() != n
-        || yfil.len() != n
-        || zfil.len() != n
-        || dlxfil.len() != n
-        || dlyfil.len() != n
-        || dlzfil.len() != n
-        || ifil.len() != n
-    {
-        return Err("Input length mismatch");
+    check_length!(m, xp, yp, zp, bx, by, bz);
+    check_length!(n, xfil, yfil, zfil, dlxfil, dlyfil, dlzfil, ifil);
+
+    // Zero output
+    bx.fill(0.0);
+    by.fill(0.0);
+    bz.fill(0.0);
+
+    // For each filament, evaluate the contribution to each observation point.
+    //
+    // The inner function is inlined, so values that are reused between iterations
+    // can be pulled to the outer scope by the compiler and do not affect performance.
+    for i in 0..n {
+        for j in 0..m {
+            // Filament
+            let fil0 = (xfil[i], yfil[i], zfil[i]); // [m] start point
+            let fil1 = (fil0.0 + dlxfil[i], fil0.1 + dlyfil[i], fil0.2 + dlzfil[i]); // [m] end point
+            let current = ifil[i];
+
+            // Observation point
+            let obs = (xp[j], yp[j], zp[j]); // [m]
+
+            // Field contributions
+            let (bxc, byc, bzc) = flux_density_linear_filament_scalar((fil0, fil1, current), obs);
+            bx[j] += bxc;
+            by[j] += byc;
+            bz[j] += bzc;
+        }
     }
 
-    for j in 0..m {
-        bx[j] = 0.0;
-        by[j] = 0.0;
-        bz[j] = 0.0;
-    }
+    Ok(())
+}
 
-    // For each filament, evaluate the contribution to each observation point
-    for i in 0..n {
-        // Get filament midpoint
-        let dlxi = dlxfil[i]; // [m]
-        let dlyi = dlyfil[i]; // [m]
-        let dlzi = dlzfil[i]; // [m]
-        let xmid = dlxi.mul_add(0.5, xfil[i]); // [m]
-        let ymid = dlyi.mul_add(0.5, yfil[i]); // [m]
-        let zmid = dlzi.mul_add(0.5, zfil[i]); // [m]
+/// Biot-Savart calculation for B-field contribution from many current filament
+/// segments to many observation points.
+///
+/// Uses filament midpoint as field source.
+///
+/// # Arguments
+///
+/// * `xyzifil`:   (m, m, A) Filament start and end coords and current
+/// * `xyzp`:     (m) Observation point coords
+///
+/// # Returns
+///
+/// * `b`:        (T) Magnetic flux density (B-field)
+pub fn flux_density_linear_filament_scalar(
+    xyzifil: ((f64, f64, f64), (f64, f64, f64), f64),
```

Comment:
Nitpick: it seems inconsistent to me that the other variants of `flux_density_linear_filament*` take `ifil` as a separate argument, but that this scalar variant groups the filament current in with the start and end points as one big tuple. I would find the arguments easier to read if this was broken into

```suggestion
    xyzfil0: (f64, f64, f64),
    xyzfil1: (f64, f64, f64),
    ifil: f64,
```

Ditto for the other `*_linear_filament_scalar` functions.

Reply:
That's a deliberate change to move toward grouping the values by association with a physical object, which makes it less confusing, and also to reduce the number of arguments. I plan to update the vector and parallel implementations to the same format in the next major change, but left that aside for this PR primarily to reduce the size of the PR.